File: PTA/PTA1013/BattleOverCities.py
from typing import List, Callable, Dict, Set
from copy import deepcopy


def dfs(start: int, graph: List[List[bool]], isVisited: List[bool]):
    """
    对图进行一次dfs
    :param isVisited: 某个边的某个临界点是否被访问
    :param start: 开始的节点
    :param graph: 图
    """
    isVisited[start] = True
    for idx, edge in enumerate(graph[start]):
        # 需要标记某个是否被访问
        if edge and not isVisited[idx]:
            dfs(idx, graph, isVisited)


def occupy_city(node: int, graph: List[List[bool]], nodeNum: int) -> int:
    """

    :param nodeNum: 一共有多少城市
    :param node: 被占领的城市
    :param graph: map
    :return: 需要修的路的数量
    """
    graph = deepcopy(graph)
    # 将边去掉
    for i in range(nodeNum):
        graph[i][node] = False
        graph[node][i] = False

    # 进行dfs的次数
    dfsTimes = -1

    nodes = set(i for i in range(1, nodeNum + 1))
    nodes -= {0, node}
    isVisited = [False for _ in range(nodeNum + 1)]
    isVisited[node] = True
    isVisited[0] = True
    while not all(isVisited):
        dfs(nodes.pop(), graph, isVisited)
        dfsTimes += 1

    return dfsTimes


def summit():
    nodeNum, edgeNum, queryNum = map(int, input().split())
    graph: List[List[bool]] = [[False] * (nodeNum + 1) for _ in range(nodeNum + 1)]
    for _ in range(edgeNum):
        a, b = map(int, input().split())
        graph[a][b] = True
        graph[b][a] = True
    # 存放结果
    result: List[int] = []

    for node in map(int, input().split()):
        result.append(occupy_city(node, graph, nodeNum))

    print("\n".join(map(str, result)))


if __name__ == '__main__':
    summit()

File: PTA/PTA1013/BattleOverCities.cpp
//
// Created by ednow on 2021/8/22.
//

#define CASEID 1013

#ifdef CASEID
#define MAIN PTA1013
#else
#define MAIN main
#endif

#include "vector"
#include "iostream"
#include "set"
//#include "unordered_set"
#include "algorithm"

//#define set unordered_set

using namespace std;

//typedef unordered_set set;

void dfs(int start, vector<vector<bool>> &graph, vector<bool> &isVisited, vector<int> &nodesToDeleted){
    isVisited[start] = true;
    nodesToDeleted.push_back(start);
    for (int i = 0; i < graph.size(); ++i) {
        if (graph[start][i] and !isVisited[i]){
            dfs(i, graph, isVisited, nodesToDeleted);
        }
    }
};

int
occupy_city(int node, vector<vector<bool>> graph, int nodeNum){
    // 进行dfs的次数
    auto dfsTimes = -1;
    for (int i = 0; i < graph.size(); ++i) {
        graph[i][node] = false;
        graph[node][i] = false;
    }
    set<int> nodes;
    for (int i = 1; i < nodeNum+1; ++i) {
        nodes.insert(i);
    }
    nodes.erase(node);
    vector<bool> isVisited(nodeNum+1, false);
    isVisited[node] = true;
    isVisited[0] = true;
    while (!nodes.empty()){
        vector<int> nodesToDeleted;
        dfs(*nodes.begin(), graph, isVisited, nodesToDeleted);
        for (int & iter : nodesToDeleted){
            nodes.erase(iter);
        }
        dfsTimes++;
    }

    return dfsTimes;
}

int MAIN(){
    int nodeNum{}, edgeNum{}, queryNum{};
    cin >> nodeNum >> edgeNum >> queryNum;
    vector<vector<bool>> graph(nodeNum + 1, vector<bool>(nodeNum + 1, false));
    while (edgeNum--){
        int a, b;
        cin >> a >> b;
        graph[a][b] = true;
        graph[b][a] = true;
    }

    vector<int> repairNums;
    repairNums.reserve(queryNum);
    while (queryNum--){
        int node;
        cin >> node;
        repairNums.push_back(occupy_city(node, graph, nodeNum));
    }

    // 输出结果
    for (int & repairNum : repairNums) {
        cout << repairNum << endl;
    }

    return 0;
};

#ifdef CASEID
#include "utils.h"

TEST(PTA, CASEID){
    DO_TEST(CASEID)
}
#endif